src/highlight_processor.c

```diff
@@ -0,0 +1,200 @@
+      order = HIGHLIGHT_ORDER_SCOREPOS;
+      break;
+
+    case SummarizeMode_WholeField:
+    case SummarizeMode_None:
+    default:
+      // Should never reach here.
+      order = -1;
+      break;
+  }
+
+  // Storage for the list of fragments
+  iovsArr = calloc(numIovArr, sizeof(*iovsArr));
+
+  FragmentList_HighlightFragments(&frags, &tags, fieldInfo->contextLen, iovsArr, numIovArr, order);
+
+  // Buffer to store concatenated fragments
+  Array bufTmp;
+  Array_Init(&bufTmp);
+
+  for (size_t ii = 0; ii < numIovArr; ++ii) {
+    Array *curIovs = iovsArr + ii;
+    struct iovec *iovs = ARRAY_GETARRAY_AS(curIovs, struct iovec *);
+    size_t numIovs = ARRAY_GETSIZE_AS(curIovs, struct iovec);
+    size_t lastSize = bufTmp.len;
+
+    for (size_t jj = 0; jj < numIovs; ++jj) {
+      Array_Write(&bufTmp, iovs[jj].iov_base, iovs[jj].iov_len);
+    }
+
+    // Duplicate spaces for the current snippet are eliminated here. We shouldn't
+    // move it to the end because the delimiter itself may contain a special kind
+    // of whitespace.
+    size_t newSize = stripDuplicateSpaces(bufTmp.data + lastSize, bufTmp.len - lastSize);
+    Array_Resize(&bufTmp, lastSize + newSize);
+    Array_Write(&bufTmp, HL_SEP_STR, sizeof(HL_SEP_STR) - 1);
+  }
+
+  // Set the string value to the contents of the array. It might be nice if we didn't
+  // need to strndup it.
+  RSFieldMap_Set(&r->fields, fieldName, RS_StringVal(strndup(bufTmp.data, bufTmp.len), bufTmp.len));
+
+  Array_Free(&bufTmp);
+  for (size_t ii = 0; ii < numIovArr; ++ii) {
+    Array_Free(iovsArr + ii);
+  }
+  free(iovsArr);
+  FragmentList_Free(&frags);
+}
+
+static int hlp_Next(ResultProcessorCtx *ctx, SearchResult *r) {
+  int rc = ResultProcessor_Next(ctx->upstream, r, 0);
+  if (rc == RS_RESULT_EOF) {
+    return rc;
+  }
+
+  RSSearchRequest *req = ctx->privdata;
+  RSByteOffsets *byteOffsets = r->md->byteOffsets;
+
+  // Assume we have highlighting here
+  for (size_t ii = 0; ii < req->fields.numFields; ++ii) {
+    if (req->fields.fields[ii].mode == SummarizeMode_None) {
+      // Ignore - this is a field for `RETURN`, not `SUMMARIZE`
+      continue;
+    }
+
+    // Otherwise, summarize this field now
+    const ReturnedField *rf = req->fields.fields + ii;
+    const char *fName = req->fields.rawFields[rf->nameIndex];
+    RSValue *fieldValue = RSFieldMap_Get(r->fields, fName);
+
+    // Check if we have the value for this field.
+    // TODO: Might this be 'RSValue_String' later on?
+    if (fieldValue == NULL || !RSValue_IsString(fieldValue)) {
+      RSFieldMap_Set(&r->fields, fName, RS_NullVal());
+      continue;
+    }
+
+    summarizeField(req, rf, fName, fieldValue, r);
+  }
+  return RS_RESULT_OK;
+}
+
+static void hlp_Free(ResultProcessor *rp) {
+  free(rp);
+}
+
+ResultProcessor *NewHighlightProcessor(ResultProcessor *parent, RSSearchRequest *req) {
+  ResultProcessor *rp = NewResultProcessor(parent, req);
+  rp->Next = hlp_Next;
+  rp->Free = hlp_Free;
```

Comment:
RTFM - if Free is NULL we call free(rp) automatically, so no need to implement it if you don't have special logic